docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+
+Bricks design explained
+=======================
+
+This is a runnable tutorial, which means you can copy its fragments 
+to an IPython notebook and run. The import section follows:
+
+    >>> import theano
+    >>> theano.config.compute_test_value = 'raise'
+    >>> from theano import tensor
+
+Please keep in mind that the code in this tutorial was written for illustration
+purposes only. Often only the common case is handled and corner cases are
+ignored.
+
+Introduction
+------------
+
+Deep learning (DL) models implementing using Theano share numerous
+building blocks which become more and more standartized over time.
+Factoring out such components is a very lucrative idea, and many people
+have fallen victims to the temptation to develop a library of DL
+building blocks. The list of rather well-known projects includes
+`PyLearn2`_, `Groundhog`_, `Lasagne`_, `Keras`_ and of course Blocks. Many people and
+research groups have smaller and more specialized frameworks.
+
```

Comment:
I would add:
At the first sight, building a DL framework is easy with tools like Theano - after all one only needs to connect a few pieces to get a first, working multilayer feed-forward net, the cornerstone of deep learning. However, in our experience a DL framework needs much more than just being able to implement small models. In particular we would like to:
1. Reuse parts of models in other models without copy-pasting code and manual model conversion. This implies hierarchical model design
2. Be able to quickly inspect what happens inside of the model, allow for ad-hoc monitoring of intermediate computations.
3. Exchange parts of models  - you want to compare a GRU layer with an LSTM one in a large machine translation network - no problem.

Many aspects of Bricks - the pieces one builds models in Blocks were added to handle our needs. We want blocks to be a library that scales and allows easy prototyping of even the most complicated models you can imagine.